src/java/edu/slu/action/ObjectAction.java

```diff
@@ -527,116 +527,167 @@ public void batchSaveFromCopy() throws UnsupportedEncodingException, IOException
+                out = response.getWriter();
+                out.write(mapper.writer().withDefaultPrettyPrinter().writeValueAsString(descendants));
+            } 
+            catch (IOException ex) {
+                Logger.getLogger(ObjectAction.class.getName()).log(Level.SEVERE, null, ex);
+            }
+        }
     }
     
     /**
-     * Servlet method to find all downstream versions of an object.
-     * If this object is the last, the return will be null.
-     * @TODO this needs to return an array of JSONObjects.
-     * @param  http_request Servlet request for relatives
-     * @throws Exception 
+     * Internal method to find all downstream versions of an object.  It should always receive a reliable object, not one from the user.
+     * If this object is the last, the return will be an empty JSONArray.  The keyObj WILL NOT be a part of the array.  
+     * @param  ls_versions All the given versions, including root, of a provided object.
+     * @param  keyObj The provided object
+     * @param  discoveredDescendants The array storing the descendants objects discovered by the recursion.
+     * @return All the objects that were deemed descendants in a JSONArray
      */
-    public void getAllDescendants(HttpServletRequest http_request) throws Exception{
-        // TODO: @theHabes, this is waiting for something clever to happen.
-        // This code is not correct at all, but pseudo-correct.
-        List<DBObject> ls_versions = getAllVersions(http_request);
-        // cubap: At this point, we have all the versions of the object (except maybe the
-        // original?) and need to filter to the ones we want.
-        // Getting the whole document is a mess, but if we get subdocuments of __rerum, 
-        // we don't need to worry as much.
-        
-        JSONArray descendants = getAllDescendants(ls_versions);
-        try {
-            response.addHeader("Access-Control-Allow-Origin", "*");
-            response.setStatus(HttpServletResponse.SC_OK);
-            out = response.getWriter();
-            out.write(mapper.writer().withDefaultPrettyPrinter().writeValueAsString(descendants));
-        } 
-        catch (IOException ex) {
-            Logger.getLogger(ObjectAction.class.getName()).log(Level.SEVERE, null, ex);
+
+    private JSONArray getAllDescendants(List<DBObject> ls_versions, JSONObject keyObj, JSONArray discoveredDescendants){
+        JSONArray nextIDarr = new JSONArray();
+        if(keyObj.getJSONObject("__rerum").getJSONObject("history").getJSONArray("next").size() > 0){
+            //essentially, do nothing.  This branch is done.
+        }
+        else{
+            //The provided object has nexts, get them to add them to known descendants then check their descendants.
+            nextIDarr = keyObj.getJSONObject("__rerum").getJSONObject("history").getJSONArray("next");
         }
+        //@cubap what should I do if I find malformed objects here?  Do we need a try-catch?
```

Comment:
The history would possibly have other versions, so it may be hard to find the next/previous. Until we have the helper for version checking, I think we should add it to the list and treat it like a leaf.